kernel/src/scan/data_skipping.rs

```diff
@@ -67,6 +106,26 @@ fn as_data_skipping_predicate(expr: &Expr) -> Option<Expr> {
             let col = format!("{}.{}", stats_col, col);
             Some(Expr::binary(op, Column(col), Literal(val.clone())))
         }
+        UnaryOperation { op, expr } => match op {
+            UnaryOperator::Not => {
+                // get the expr as a skipping predicate, then invert it
+                as_data_skipping_predicate(expr).map(Expr::not)
```

Comment:
aside: I notice there are a lot of `Expr::` qualifiers in this method, even tho the method starts with `use Expr::*` ?

Reply:
We could also just add match arms for the various `Not`, and recurse on a rewritten form, perhaps simpler/cleaner?

<details>

```rust
fn as_inverted_data_skipping_predicate(expr: Expr) -> Option<Expr> {
    match expr {
        UnaryOperation { op, expr } => match op {
            UnaryOperator::Not => as_data_skipping_predicate(expr),
            UnaryOperator::IsNull => { ... IS NOT NULL case ... },
         },
         BinaryOperation { op, left, right } => {
             let expr = binary(op.invert()?, left, right);
             as_data_skipping_predicate(&expr)
         }
         VariadicOperation { op, exprs } => {
             let expr = variadic(op.invert()?, exprs.into_iter().map(Expr::not);
             as_data_skipping_predicate(&expr)
         }
    }
}

fn as_data_skipping_predicate(expr: &Expr) -> Option<Expr> {
      ...
    match expr {
        UnaryOperation { op, expr } => match op {
            UnaryOperator::Not => as_inverted_data_skipping_predicate(expr),
            UnaryOperator::IsNull => { ... IS NULL case ... },
        }
        ... existing code unchanged ...
```
</details>

Reply:
So, I think it's possibly useful to push down and eliminate `NOT`, but I don't think it removes the need for "null vs. false". Just because we invert `IS NULL` to `IS NOT NULL`, we still need the same checks on `nullCount` to see if it could be null or not, and then we'd just invert that result, and have the same issues with `nulls`.

Also, to "invert" the result, we'll still need to have the engine support `NOT` as an expression (we can't poke at the answer so we can't just "invert" it).  Am I missing something here?

Pushdown could often involve fewer expressions to evaluate though, so probably worth it for that.

Reply:
> aside: I notice there are a lot of `Expr::` qualifiers in this method, even tho the method starts with `use Expr::*` ?

Things like `Expr::le` aren't imported by that, so still need to be specified. We could change something like:
```rust
Expr::le(Column(col.clone()), Literal(val.clone())),
```

to

```rust
Column(col.clone()).le(Literal(val.clone())),
```

But not sure it's worth it

Reply:
> Also, to "invert" the result, we'll still need to have the engine support `NOT` as an expression (we can't poke at the answer so we can't just "invert" it). Am I missing something here?

I realize we could just note on the filter itself that the result should be inverted, and then could do so in the visitor. I need to think about if that always works, but perhaps then we could eliminate `NOT` as an expression completely.

Reply:
Dang, this is now "outdated". Regardless, this is an interesting discussion. I thought a bit more, and yes, I think `NULL` is the only case where we'd need to do something special. We can't just get rid of NOT (without an `IsNotNull`)  and invert at the end (which you probably already realized) because of an expression like:
```
AND(NOT IS NULL, "number" >  2)
```

But other than that, we can just transform and not need not (heh).

`IsNotNull` seems pretty simple as an expression type, although I guess it would still require the engine to have some notion of "not".

Would you be okay with merging like this for now, and moving this to an issue for "future work"?